platforms/ide/problems/src/main/kotlin/org/gradle/problems/internal/DefaultProblemsReportCreator.kt

```diff
@@ -0,0 +1,125 @@
+import org.gradle.internal.concurrent.ExecutorFactory
+import org.gradle.internal.configuration.problems.CommonReport
+import org.gradle.internal.configuration.problems.FailureDecorator
+import org.gradle.internal.configuration.problems.StructuredMessage
+import org.gradle.internal.configuration.problems.writeError
+import org.gradle.internal.configuration.problems.writeStructuredMessage
+import org.gradle.internal.logging.ConsoleRenderer
+import org.gradle.internal.operations.OperationIdentifier
+import org.gradle.internal.problems.failure.FailureFactory
+import org.gradle.problems.buildtree.ProblemReporter
+import java.io.File
+
+val logger: Logger = Logging.getLogger(DefaultProblemsReportCreator::class.java)
+
+class DefaultProblemsReportCreator(
+    executorFactory: ExecutorFactory?,
+    temporaryFileProvider: TemporaryFileProvider?,
+    internalOptions: InternalOptions?,
+    startParameter: StartParameterInternal,
+    private val failureFactory: FailureFactory,
+    private val buildNameProvider: BuildNameProvider
+) : ProblemReportCreator {
+
+    private val report = CommonReport(executorFactory!!, temporaryFileProvider!!, internalOptions!!, "problem-report")
+    private val taskNames: List<String> = startParameter.taskNames
+    private var problemCount = 0
+
+    private val failureDecorator = FailureDecorator()
+
+    override fun getId(): String {
+        return "DefaultProblemsReportCreator"
+    }
+
+    override fun report(reportDir: File, validationFailures: ProblemReporter.ProblemConsumer) {
+        report.writeReportFileTo(reportDir, object : JsonSource {
+            override fun writeToJson(jsonWriter: JsonWriter) {
+                with(jsonWriter) {
+                    property("reportContext", "problems report")
+                    property("totalProblemCount", problemCount.toString())
+                    buildNameProvider.buildName()?.let { property("buildName", it) }
+                    property("requestedTasks", taskNames.joinToString(" "))
+                    property("cacheAction", "cacheAction")
+                    property("cacheActionDescription") {
+                        writeStructuredMessage(StructuredMessage.Builder().text("").build())
+                    }
+                    property("documentationLink", DocumentationRegistry().getDocumentationFor("problem-report"))
+                    property("documentationLinkCaption", "Problem Report")
+                }
+            }
+        })?.let {
+            val url = ConsoleRenderer().asClickableFileUrl(it)
+            logger.warn("Problems report is available at: $url")
+        }
+    }
+
+    override fun emit(problem: Problem, id: OperationIdentifier?) {
+        problemCount++
+        report.onProblem(object : JsonSource {
+            override fun writeToJson(jsonWriter: JsonWriter) {
+                with(jsonWriter) {
+                    jsonObject {
+                        property("trace") {
```

Comment:
💭  I am starting to doubt the current approach is maintainable. It's becoming rather fragile.

Previously, there was a single place where we've written the data in the JSON format, which was effectively the contract expected by the JavaScript part. Now, this is at least in two different places.

Reply:
As of now there will be 2 contracts.
Therefore it can obviously not stay in the same place it was.
Maybe a collection of extension methods per "contract" could avoid the separation?